src/libraries/System.Text.RegularExpressions/gen/RegexAnalyzer.cs

```diff
@@ -0,0 +1,416 @@
+namespace System.Text.RegularExpressions.Generator
+{
+    /// <summary>
+    /// Roslyn analyzer that searches for invocations of the Regex constructors, or the
+    /// Regex static methods and analyzes if the callsite could be using the Regex Generator instead.
+    /// If so, it will emit an informational diagnostic to suggest use the Regex Generator.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class RegexAnalyzer : DiagnosticAnalyzer
+    {
+        // private members
+        private const string RegexTypeName = "System.Text.RegularExpressions.Regex";
+        private const string RegexGeneratorTypeName = "System.Text.RegularExpressions.RegexGeneratorAttribute";
+        private const string TimeSpanTypeName = "System.TimeSpan";
+        private const string TimeoutTypeName = "System.Threading.Timeout";
+
+        // internal members
+        internal const string PatternIndexName = "PatternIndex";
+        internal const string RegexOptionsIndexName = "RegexOptionsIndex";
+        internal const string RegexTimeoutIndexName = "RegexTimeoutIndex";
+        internal const string RegexTimeoutName = "RegexTimeout";
+        internal const string DiagnosticId = "SYSLIB1046";
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticDescriptors.UseRegexSourceGeneration);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            // Register analysis of calls to the Regex constructors
+            context.RegisterOperationAction(AnalyzeObjectCreation, OperationKind.ObjectCreation);
+
+            // Register analysis of calls to Regex static methods
+            context.RegisterOperationAction(AnalyzeInvocation, OperationKind.Invocation);
+        }
+
+        /// <summary>
+        /// Analyzes an invocation expression to see if the invocation is a call to one of the Regex static methods,
+        /// and checks if they could be using the source generator instead.
+        /// </summary>
+        /// <param name="context">The operation context representing the invocation.</param>
+        private void AnalyzeInvocation(OperationAnalysisContext context)
+        {
+            // Ensure source generator is supported.
+            if (!ProjectSupportsRegexSourceGenerator(context.Operation))
+            {
+                return;
+            }
+
+            // Ensure the invocation is a Regex static method.
+            IInvocationOperation invocationOperation = (IInvocationOperation)context.Operation;
+            IMethodSymbol method = invocationOperation.TargetMethod;
+            if (!method.IsStatic || !SymbolEqualityComparer.Default.Equals(method.ContainingType, context.Compilation.GetTypeByMetadataName(RegexTypeName)))
```

Comment:
Per my earlier comment, as part of that startup work prior to registration of AnalyzeInvocation, you'd grab this symbol into a local, and then here you wouldn't be looking it up in the compilation each time, as that adds measurable expense.